dashboard/app/helpers/aichat_sagemaker_helper.rb

```diff
@@ -1,62 +1,126 @@
 module AichatSagemakerHelper
   ASSISTANT = "assistant"
   USER = "user"
+  SYSTEM = "system"
+  KAREN_PRETEXT = "Edit the following text for spelling and grammar mistakes: "
   INSTRUCTIONS_BEGIN_TOKEN = "[INST]"
   INSTRUCTIONS_END_TOKEN = "[/INST]"
   SENTENCE_BEGIN_TOKEN = "<s>"
   SENTENCE_END_TOKEN = "</s>"
+  CHAT_ML_BEGIN_TOKEN = "<|im_start|>"
+  CHAT_ML_END_TOKEN = "<|im_end|>"
+  NEWLINE = "\n"
   MAX_NEW_TOKENS = 512
   TOP_P = 0.9
+  PIRATE_CUSTOM_STOPPING_STRINGS = ["},"]
+  FINE_TUNED_MODELS = {
+    ARITHMO: "gen-ai-arithmo2-mistral-7b",
+    PIRATE: "gen-ai-mistral-pirate-7b",
+    KAREN: "gen-ai-karen-creative-mistral-7b",
+    BIOMISTRAL: "gen-ai-biomistral-7b"
+  }
 
   def self.create_sagemaker_client
     Aws::SageMakerRuntime::Client.new
   end
 
-  # The instruction-tuned version of Mistral accepts formatted instructions where conversation roles
-  # must start with a user prompt and alternate between user and assistant.
-  # Mistral-7B-Instruction LLM instruction format doc at https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.1.
   def self.format_inputs_for_sagemaker_request(aichat_params, stored_messages, new_message)
-    all_messages = [*stored_messages, new_message]
-    inputs = aichat_params[:systemPrompt] + " "
-    inputs += aichat_params[:retrievalContexts].join(" ") if aichat_params[:retrievalContexts]
-    inputs = SENTENCE_BEGIN_TOKEN + wrap_as_instructions(inputs)
-    all_messages.each do |msg|
-      if msg[:role] == USER
-        inputs += wrap_as_instructions(msg[:chatMessageText])
-      elsif msg[:role] == ASSISTANT
-        # Note that each assistant message in the conversation history is followed by
-        # the end-of-sentence token but a begin-of-sentence token is not required.
-        inputs += msg[:chatMessageText] + SENTENCE_END_TOKEN
+    selected_model_id = aichat_params[:selectedModelId]
+    # Add system prompt and retrieval contexts if available to inputs as part of instructions that will be sent to model.
+    inputs = ""
+    inputs = aichat_params[:systemPrompt] + " " unless aichat_params[:systemPrompt].empty?
+    inputs << aichat_params[:retrievalContexts].join(" ") if aichat_params[:retrievalContexts]
+    if selected_model_id == FINE_TUNED_MODELS[:ARITHMO]
```

Comment:
I'd recommend breaking these out into separate files to handle processing for each specific model. Especially as we add more models, maintaining a long if-else chain here could get unwieldy. For example, maybe something like
```
def get_model_processor(model)
  case model
  when 'gen-ai-arithmo2-mistral-7b'
    Arithmo2Processor
  when 'gen-ai-mistral-7b-inst-v01'
    MistralInstProcessor
  # ... etc
  else
    # maybe a default or throw an error?
  end
end

# Then you can use the given processor to format the input, output, and get custom stop strings

processor = get_model_processor(selectedModelId)

inputs = processor.format_inputs(systemPrompt, retrievalContexts, stored_messages, new_message)

...

parameters: {
  ...,
  stop: processor.get_stop_strings,
}

...

response = processor.format_output(generated_text)
```

where a single processor just needs to implement the specific functions (i.e. `format_inputs`, `format_outputs`, `get_stop_strings` or something)

That way the logic for each model stays contained to its specific file, and this file can stay relatively simple.

Edit: also, if certain input/output processing is common across some models, we could have some hierarchy to the processor classes. Ex - maybe all of these classes extend from some base `MistralProcessor` class so that BioMistral, Mistral, and Pirate Mistral can use the same input formatting function. Then for example later on if we add Llama or Claude or something, we can create a new base processor class.

Reply:
Great idea! Will update.

Reply:
@sanchitmalhotra126  - I ended up creating a base class `MistralProcessor`, and extending this class for all the other processors including `ArithmoProcessor` and `KarenProcessor` (for the `get_stop_strings` method).

A couple things I learned:
- I needed to `require_relative` first `mistral_processor`.
- I was unable to use:
```
  Dir[File.expand_path('./ai_model_processors/*.rb', __FILE__)].each {|file| require file}
```
Not sure if the 2nd item is related to 1st but just wanting to note.

Anyways - lmk what you think! Thanks so much - I think it really helps clean up the code.